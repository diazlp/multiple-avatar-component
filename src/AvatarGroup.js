import PropTypes from 'prop-types';
import Avatar from './Avatar'

const AvatarGroup = ({ maxLength, size }) => {
  //default names length -> 10
  const names = ['Tomi Budi', 'Diaz Linggaputra', 'John Doe', 'Jane Samsonite', 'Johnson Johnson', 'David Lee', 'Sarah Brown', 'Reinth Byakura', 'Shia Labouf', "Lewis Hadnott"];

  const visibleNames = names.slice(0, maxLength)
  const hiddenNamesCount = names.length - maxLength

  return (
    <div data-testid="avatar-group" className="avatar-group flex flex-nowrap items-center -space-x-6">
      {
        visibleNames.map((name, index) => (
          <Avatar key={index} name={name} size={size} dataTestId={`avatar-${index}`} />
        ))
      }
      {
        hiddenNamesCount > 0 && (
          <Avatar name={hiddenNamesCount} size={size} dataTestId="avatar-hidden-count" />
        )
      }
    </div>
  );
};

AvatarGroup.propTypes = {
  maxLength: PropTypes.number.isRequired,
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
};

AvatarGroup.defaultProps = {
  maxLength: 1,
  size: 'md',
};

export default AvatarGroup;
