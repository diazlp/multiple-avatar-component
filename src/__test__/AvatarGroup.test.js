import { render, screen, cleanup } from '@testing-library/react';
import AvatarGroup from '../AvatarGroup';

beforeEach(() => {
  cleanup();
});

describe('AvatarGroup Component Testing', () => {
  it('renders default component when passed no props', () => {
    render(<AvatarGroup />);
    const avatarGroup = screen.getByTestId('avatar-group');
    const visibleName = screen.getByText('TB');
    const hiddenNamesCount = screen.getByText('+9');

    expect(avatarGroup).toBeInTheDocument();
    expect(visibleName).toBeInTheDocument();
    expect(hiddenNamesCount).toBeInTheDocument();
  });

  it('renders default size props when its not being passed', () => {
    render(<AvatarGroup maxLength={3} />);
    const avatarGroup = screen.getByTestId('avatar-group');
    const avatars = screen.getAllByTestId(/avatar-\d+/);

    expect(avatarGroup).toBeInTheDocument();
    expect(avatars).toHaveLength(3);

    avatars.forEach(avatar => {
      expect(avatar).toHaveClass('w-12 h-12 text-base');
    });
  });

  it('renders default maxLength props when its not being passed', () => {
    render(<AvatarGroup size={'lg'} />);
    const avatarGroup = screen.getByTestId('avatar-group');
    const avatars = screen.getAllByTestId(/avatar-\d+/);

    expect(avatarGroup).toBeInTheDocument();
    expect(avatars).toHaveLength(1);

    avatars.forEach(avatar => {
      expect(avatar).toHaveClass('w-16 h-16 text-lg');
    });
  });


  it('applies the correct avatar class if given xs as size props', () => {
    render(<AvatarGroup size={'xs'} />);
    const avatars = screen.getAllByTestId(/avatar-\d+/);

    avatars.forEach(avatar => {
      expect(avatar).toHaveClass('w-8 h-8 text-xs');
    });
  });

  it('applies the correct avatar class if given sm as size props', () => {
    render(<AvatarGroup size={'sm'} />);
    const avatars = screen.getAllByTestId(/avatar-\d+/);

    avatars.forEach(avatar => {
      expect(avatar).toHaveClass('w-10 h-10 text-sm');
    });
  });

  it('applies the correct avatar class if given md as size props', () => {
    render(<AvatarGroup size={'md'} />);
    const avatars = screen.getAllByTestId(/avatar-\d+/);

    avatars.forEach(avatar => {
      expect(avatar).toHaveClass('w-12 h-12 text-base');
    });
  });

  it('applies the correct avatar class if given lg as size props', () => {
    render(<AvatarGroup size={'lg'} />);
    const avatars = screen.getAllByTestId(/avatar-\d+/);

    avatars.forEach(avatar => {
      expect(avatar).toHaveClass('w-16 h-16 text-lg');
    });
  });

  it('renders 10 avatar if 10 is given as maxLength props', () => {
    render(<AvatarGroup maxLength={10} size={'lg'} />);
    const avatars = screen.getAllByTestId(/avatar-\d+/);

    expect(avatars).toHaveLength(10);
  });

  it('matches component snapshot', () => {
    const { container } = render(<AvatarGroup maxLength={9} size={'lg'} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});