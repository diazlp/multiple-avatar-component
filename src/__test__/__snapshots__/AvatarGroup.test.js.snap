// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup Component Testing matches component snapshot 1`] = `
<div
  class="avatar-group flex flex-nowrap items-center -space-x-6"
  data-testid="avatar-group"
>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-0"
  >
    <span
      class="text-white font-bold"
    >
      TB
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-1"
  >
    <span
      class="text-white font-bold"
    >
      DL
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-2"
  >
    <span
      class="text-white font-bold"
    >
      JD
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-3"
  >
    <span
      class="text-white font-bold"
    >
      JS
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-4"
  >
    <span
      class="text-white font-bold"
    >
      JJ
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-5"
  >
    <span
      class="text-white font-bold"
    >
      DL
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-6"
  >
    <span
      class="text-white font-bold"
    >
      SB
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-7"
  >
    <span
      class="text-white font-bold"
    >
      RB
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-8"
  >
    <span
      class="text-white font-bold"
    >
      SL
    </span>
  </div>
  <div
    class="avatar bg-gray-500 w-16 h-16 text-lg flex items-center justify-center"
    data-testid="avatar-hidden-count"
  >
    <span
      class="text-white font-bold"
    >
      +1
    </span>
  </div>
</div>
`;
